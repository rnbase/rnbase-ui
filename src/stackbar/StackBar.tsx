import React from 'react';
import { StyleProp, StyleSheet, Text, TextStyle, View, ViewProps, ViewStyle } from 'react-native';

import { Theme, Themed, WithThemeProps, withTheme } from '../theming';
import { getArrayColor, getRadius } from '../helpers';

function getTopItems(items: DataProps[], count: number) {
  return [...items.entries()]
    .sort((a, b) => b[1].value - a[1].value)
    .splice(0, count)
    .sort((a, b) => a[0] - b[0])
    .map(item => item[1]);
}

interface DataProps {
  value: number;
  color?: string;
  label?: string;
}

interface StackBarProps extends ViewProps {
  size?: number;
  cornerRadius?: number | string;
  data: DataProps[];
  separatorWidth?: number;
  showLabels?: number;
  renderLabel?: (item: DataProps, index: number) => void;
  style?: StyleProp<ViewStyle>;
  legendStyle?: StyleProp<ViewStyle>;
  legendLabelStyle?: StyleProp<ViewStyle>;
  legendSwatchStyle?: StyleProp<ViewStyle>;
  legendTextStyle?: StyleProp<TextStyle>;
}

const StackBar: React.FC<Themed<typeof createStyleSheet, StackBarProps>> = ({
  theme: { styles },
  data,
  size = 10,
  cornerRadius = '50%',
  separatorWidth = 1,
  showLabels = data.length,
  renderLabel: propRenderLabel,
  style,
  legendStyle,
  legendLabelStyle,
  legendSwatchStyle,
  legendTextStyle,
  ...props
}) => {
  const count = data.length;
  const items = data.map((item, index) => ({
    ...item,
    color: item.color || getArrayColor(index, count),
  }));

  const renderItem = ({ value, color }: DataProps, index: number) => {
    const itemStyles = {
      height: '100%',
      flexGrow: value,
      backgroundColor: color,
      marginLeft: index > 0 ? separatorWidth : 0,
    };

    return <View key={index} style={itemStyles} />;
  };

  const renderLabel = ({ color, label }: DataProps, index: number) => {
    const swatchStyles = [{ backgroundColor: color }, styles.legendSwatch, legendSwatchStyle];

    return !label ? null : (
      <View key={index} style={[styles.legendLabel, legendLabelStyle]}>
        <View style={swatchStyles} />
        <Text style={[styles.legendText, legendTextStyle]} numberOfLines={1}>
          {label}
        </Text>
      </View>
    );
  };

  const rootStyles = [
    {
      height: size,
      borderRadius: getRadius(cornerRadius, size),
    },
    styles.root,
  ];

  return (
    <View {...props} style={style}>
      <View style={rootStyles}>{items.map(renderItem)}</View>
      {showLabels > 0 && (
        <View style={[styles.legend, legendStyle]}>
          {getTopItems(items, showLabels).map(propRenderLabel || renderLabel)}
        </View>
      )}
    </View>
  );
};

const createStyleSheet = ({ colors, fonts }: Theme) =>
  StyleSheet.create({
    root: {
      overflow: 'hidden',
      flexDirection: 'row',
    },
    legend: {
      marginTop: 8,
      flexWrap: 'wrap',
      flexDirection: 'row',
    },
    legendLabel: {
      marginRight: 10,
      flexDirection: 'row',
      alignItems: 'center',
    },
    legendSwatch: {
      width: 10,
      height: 10,
      marginRight: 3,
      borderRadius: 5,
    },
    legendText: {
      ...fonts.normal,
      fontSize: 12,
      color: colors.black,
    },
  });

export type Props = StackBarProps & WithThemeProps;

export default withTheme(StackBar, createStyleSheet, 'StackBar');
