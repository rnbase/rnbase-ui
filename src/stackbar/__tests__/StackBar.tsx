import React from 'react';

// Note: test renderer must be required after react-native.
import TestRenderer from 'react-test-renderer';

/**
 * Under test
 */
import StackBar, { Props } from '../StackBar';

const data = [
  {
    value: 1,
    color: 'red',
    label: 'One',
  },
  {
    value: 2,
    color: 'green',
    label: 'Two',
  },
  {
    value: 3,
    color: 'blue',
    label: 'Three',
  },
];

const createElement = (props: Partial<Props>) => <StackBar data={data} {...props} />;

const createRenderer = (props: Partial<Props>) => TestRenderer.create(createElement(props));

it('should render normally', () => {
  expect(createRenderer({})).toMatchSnapshot();
});

it('should render without colors and labels', () => {
  const props = { showLegend: false, data: [{ value: 1 }, { value: 2 }, { value: 3 }] };

  expect(createRenderer(props)).toMatchSnapshot();
});
