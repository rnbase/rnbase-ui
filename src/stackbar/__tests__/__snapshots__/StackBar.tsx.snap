// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render normally 1`] = `
<View>
  <View
    style={
      Array [
        Object {
          "borderRadius": 5,
          "height": 10,
        },
        Object {
          "flexDirection": "row",
          "overflow": "hidden",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "red",
          "flexGrow": 1,
          "height": "100%",
          "marginLeft": 0,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "green",
          "flexGrow": 2,
          "height": "100%",
          "marginLeft": 1,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "blue",
          "flexGrow": 3,
          "height": "100%",
          "marginLeft": 1,
        }
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "flexWrap": "wrap",
          "marginTop": 8,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginRight": 10,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "red",
            },
            Object {
              "borderRadius": 5,
              "height": 10,
              "marginRight": 3,
              "width": 10,
            },
            undefined,
          ]
        }
      />
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#000000",
              "fontFamily": "System",
              "fontSize": 12,
              "fontWeight": "400",
              "includeFontPadding": false,
              "textAlignVertical": "center",
            },
            undefined,
          ]
        }
      >
        One
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginRight": 10,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "green",
            },
            Object {
              "borderRadius": 5,
              "height": 10,
              "marginRight": 3,
              "width": 10,
            },
            undefined,
          ]
        }
      />
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#000000",
              "fontFamily": "System",
              "fontSize": 12,
              "fontWeight": "400",
              "includeFontPadding": false,
              "textAlignVertical": "center",
            },
            undefined,
          ]
        }
      >
        Two
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginRight": 10,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "blue",
            },
            Object {
              "borderRadius": 5,
              "height": 10,
              "marginRight": 3,
              "width": 10,
            },
            undefined,
          ]
        }
      />
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#000000",
              "fontFamily": "System",
              "fontSize": 12,
              "fontWeight": "400",
              "includeFontPadding": false,
              "textAlignVertical": "center",
            },
            undefined,
          ]
        }
      >
        Three
      </Text>
    </View>
  </View>
</View>
`;

exports[`should render without colors and labels 1`] = `
<View>
  <View
    style={
      Array [
        Object {
          "borderRadius": 5,
          "height": 10,
        },
        Object {
          "flexDirection": "row",
          "overflow": "hidden",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "hsl(360, 65%, 65%)",
          "flexGrow": 1,
          "height": "100%",
          "marginLeft": 0,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "hsl(240, 65%, 65%)",
          "flexGrow": 2,
          "height": "100%",
          "marginLeft": 1,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "hsl(120, 65%, 65%)",
          "flexGrow": 3,
          "height": "100%",
          "marginLeft": 1,
        }
      }
    />
  </View>
</View>
`;
